EDM-854: Use go-templates for fleet template parameters (#700)
flightctl/flightctl: docs/user/managing-fleets.md

For example, you could specify in a fleet's device template that all devices in the fleet shall run the OS image `quay.io/flightctl/rhel:9.5`. The Flight Control service would then roll out this specification to all devices in the fleet and the Flight Control agents would update the devices accordingly. The same would apply to the other specification items described in [Managing Devices](managing-devices.md).

> However, it would be impractical if *all* of a fleet's devices had to have the *exact same specification*. Flight Control therefore allows templates to contain placeholders that get filled in based on a device's name or label values. The syntax for these placeholders matches that of [Go templates](https://pkg.go.dev/text/template). You may reference anything under a device's metadata such as `{{ index .Metadata.Labels \"key\" }}` or {{ .Metadata.Name }}`.

Author: Fixed

Here are some examples what you can do with placeholders in device templates:

> * You can label devices by their deployment stage (say, `stage: testing` and `stage: production`) and then use the label with the key `stage` as placeholder when referencing the OS image to use (say, `quay.io/myorg/myimage:latest-{{ index .Metadata.Labels \"stage\" }}`) or when referencing a folder with configuration in a Git repository.

Author: If a label doesn't exist then:
{{index .Metadata.Labels "key"}} defaults to ""
{{.Metadata.Labels.key}} defaults to "<no value>"

> * You can label devices by their deployment stage (say, `stage: testing` and `stage: production`) and then use the label with the key `stage` as placeholder when referencing the OS image to use (say, `quay.io/myorg/myimage:latest-{{ index .Metadata.Labels \"stage\" }}`) or when referencing a folder with configuration in a Git repository.

Reviewer: So both have the same behaviour? In that case, I personally find the latter easier.

> * You can label devices by their deployment stage (say, `stage: testing` and `stage: production`) and then use the label with the key `stage` as placeholder when referencing the OS image to use (say, `quay.io/myorg/myimage:latest-{{ index .Metadata.Labels \"stage\" }}`) or when referencing a folder with configuration in a Git repository.

Author: {{index .Metadata.Labels "key"}} defaults to ""
{{.Metadata.Labels.key}} defaults to "<no value>"

> * You can label devices by their deployment stage (say, `stage: testing` and `stage: production`) and then use the label with the key `stage` as placeholder when referencing the OS image to use (say, `quay.io/myorg/myimage:latest-{{ index .Metadata.Labels \"stage\" }}`) or when referencing a folder with configuration in a Git repository.

Author: Sorry, github messed up my comment, please see above :)

> * You can label devices by their deployment stage (say, `stage: testing` and `stage: production`) and then use the label with the key `stage` as placeholder when referencing the OS image to use (say, `quay.io/myorg/myimage:latest-{{ index .Metadata.Labels \"stage\" }}`) or when referencing a folder with configuration in a Git repository.

Reviewer: Got it. So if we change to missingkey=error as discussed below the {{.Metadata.Labels.key}} would return an error instead.

For example, you could specify in a fleet's device template that all devices in the fleet shall run the OS image `quay.io/flightctl/rhel:9.5`. The Flight Control service would then roll out this specification to all devices in the fleet and the Flight Control agents would update the devices accordingly. The same would apply to the other specification items described in [Managing Devices](managing-devices.md).

> However, it would be impractical if *all* of a fleet's devices had to have the *exact same specification*. Flight Control therefore allows templates to contain placeholders that get filled in based on a device's name or label values. The syntax for these placeholders matches that of [Go templates](https://pkg.go.dev/text/template), but you may only use simple text or actions (no conditionals or loops, for example). You may reference anything under a device's metadata such as `{{ .Metadata.Labels.key }}` or `{{ .Metadata.Name }}`.

Reviewer: I it surprising to users that they'd have to use the variable names as in our Go structs instead of those in the YAML files (IOW, they need to use `.Metadata.Name` instead of `.metadata.name`)? in these examples it's just the initial capital letter, but in theory the Go field and YAML field could be named very differently.

Author: Good point.  Need to think about how to translate that.

We also provide some helper functions:

> * `toUpper`: Change to upper case. For example, `{{ toUpper .Metadata.Name }}`.

Reviewer: Should we support and prefer the pipeline syntax?

If there are multiple actions, the pipeline syntax feels cleaner and more intuitive, e.g.:
```
{{ .Metadata.Labels.stage | default "testing" | lower | replace " " "-" }}
```

> * `toUpper`: Change to lower case. For example, `{{ toLower .Metadata.Labels.key }}`.

Reviewer 2: ```suggestion
* `toLower`: Change to lower case. For example, `{{ toLower .Metadata.Labels.key }}`.
```

Author: Fixed

> * `replace`: Replace a substring with another string. For example, `{{ replace .Metadata.Labels.key \"old\" \"new\" }}`.

Reviewer: ```suggestion
* `replace`: Replace all occurrences of substring with another string. For example, `{{ replace .Metadata.Labels.key \"old\" \"new\" }}`.
```

* `getOrDefault`: Return a default value if accessing a missing label. For example, `{{ getOrDefault .Metadata.Labels \"key\" \"default\" }}`.
* `getOrError`: Fail if accessing a missing label. For example, `{{ getOrError .Metadata.Labels \"key\" }}`.

> Here are some examples what you can do with placeholders in device templates:

Reviewer 2: ```suggestion
Here are some examples of what you can do with placeholders in device templates:
```

Author: Fixed
